lib/ecto/query/planner.ex

```diff
@@ -642,21 +659,40 @@ defmodule Ecto.Query.Planner do
     end
   end
 
-  defp expr_to_cache(%BooleanExpr{op: op, expr: expr}), do: {op, expr}
+  defp expr_to_cache(%BooleanExpr{op: op, expr: expr, subqueries: []}), do: {op, expr}
+  defp expr_to_cache(%BooleanExpr{op: op, expr: expr, subqueries: subqueries}) do
+    # Alternate implementation could be replace {:subquery, i} expression in expr.
+    # Current strategy appends [{:subquery, i, cache}], where cache is the cache key for this subquery.
+    {op, expr, subqueries |> Enum.with_index() |> Enum.map(fn {%{cache: cache}, i} -> {:subquery, i, cache} end)}
```

Comment:
I don't think we need an index. :)

```suggestion
    {op, expr, Enum.map(subqueries, fn %{cache: cache} -> {:subquery, cache} end)}
```